crypto/ec/ecx_meth.c

```diff
@@ -129,18 +242,21 @@ static int ecx_pub_decode(EVP_PKEY *pkey, X509_PUBKEY *pubkey)
 
     if (!X509_PUBKEY_get0_param(NULL, &p, &pklen, &palg, pubkey))
         return 0;
-    return ecx_key_op(pkey, pkey->ameth->pkey_id, palg, p, pklen,
-                      X25519_PUBLIC);
+    return ecx_key_op(NULL, pkey, pkey->ameth->pkey_id, palg, p, pklen,
+                      KEY_OP_PUBLIC);
 }
 
 static int ecx_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
 {
-    const X25519_KEY *akey = a->pkey.ptr;
-    const X25519_KEY *bkey = b->pkey.ptr;
+    unsigned char *pubkeya, *pubkeyb;
 
-    if (akey == NULL || bkey == NULL)
+    if (a->pkey.ptr == NULL || b->pkey.ptr == NULL)
         return -2;
-    return !CRYPTO_memcmp(akey->pubkey, bkey->pubkey, X25519_KEYLEN);
+
+    GETPUBKEY(pubkeya, a);
+    GETPUBKEY(pubkeyb, b);
+
+    return !CRYPTO_memcmp(pubkeya, pubkeyb, KEYLEN(a));
```

Comment:
shouldn't we compare key type first?
And I think the "==0" is more clear than "!"

Reply:
We are guaranteed to have keys of the same type here. That check is done for us in EVP_PKEY_cmp. I changed it to "== 0"